runtime/lua/vim/treesitter/query.lua

```diff
@@ -778,15 +934,86 @@ end
+---@deprecated
 function Query:iter_matches(node, source, start, stop, opts)
+  vim.deprecate('iter_matches', 'iter_matches2', '0.12')
+  if type(source) == 'number' and source == 0 then
+    source = api.nvim_get_current_buf()
+  end
+
+  start, stop = value_or_node_range(start, stop, node)
+
+  -- Tell _rawquery to return a match table of capture_id -> TSNode
+  -- opts._use_deprecated_and_broken_api = true
+
+  local raw_iter = node:_rawquery(self.query, false, start, stop, opts) ---@type fun(): integer, TSMatch
+  local function iter()
+    local pattern, match = raw_iter()
+    local metadata = {}
+
+    if match ~= nil then
+      local active = self:match_preds(match, pattern, source)
+      if not active then
+        return iter() -- tail call: try next match
+      end
+
+      self:apply_directives(match, pattern, source, metadata)
+    end
+
+    -- Convert the match table into the old buggy version for backward
+    -- compatibility. This is slow. Plugin authors, if you're reading this, stop
+    -- using iter_matches!
+    local old_match = {} ---@type table<integer, TSNode>
+    for k, v in pairs(match or {}) do
+      old_match[k] = v[#v]
+    end
+    return pattern, old_match, metadata
+  end
+  return iter
+end
+
+--- Iterates the matches of self on a given range.
+---
+--- Iterate over all matches within a {node}. The arguments are the same as
+--- for |Query:iter_captures()| but the iterated values are different:
+--- an (1-based) index of the pattern in the query, a table mapping
+--- capture indices to nodes, and metadata from any directives processing the match.
+--- If the query has more than one pattern, the capture table might be sparse
+--- and e.g. `pairs()` method should be used over `ipairs`.
+--- Here is an example iterating over all captures in every match:
+---
+--- ```lua
+--- for pattern, match, metadata in cquery:iter_matches(tree:root(), bufnr, first, last) do
+---   for id, nodes in pairs(match) do
+---     local name = query.captures[id]
+---     for _, node in ipairs(nodes) do
+---       -- `node` was captured by the `name` capture in the match
+---
+---       local node_data = metadata[id] -- Node level metadata
+---       ... use the info here ...
+---     end
+---   end
+--- end
+--- ```
+---
+---
+---@param node TSNode under which the search will occur
+---@param source (integer|string) Source buffer or string to search
+---@param start? integer Starting line for the search. Defaults to `node:start()`.
+---@param stop? integer Stopping line for the search (end-exclusive). Defaults to `node:end_()`.
+---@param opts? table Optional keyword arguments:
+---   - max_start_depth (integer) if non-zero, sets the maximum start depth
+---     for each match. This is used to prevent traversing too deep into a tree.
+---
+---@return (fun(): integer, table<integer, TSNode[]>, table): pattern id, match, metadata
+function Query:iter_matches2(node, source, start, stop, opts)
```

Comment:
Thanks for the efforts in compatibility APIs and soft-deprecation. Eventually (after a couple iterations of deprecation and removals) we would want to settle down with `iter_matches`.

To avoid changing it back to `iter_matches`, I suggest, instead of introducing a new method with different name (which might be deprecated again), introducing an option to `iter_matches(..., opts)`. Something like `multiple_nodes` or *`correct`*. One downside would be complicating the documentation.

Reply:
That's the best option, I feel (pun intended); since the option would be temporary, the documentation is not a big issue. This also makes the name less of a critical issue; maybe `list` or `match_list` suffices.

Reply:
A counterpoint in favor of a new function instead of an option: it makes it much more unlikely that new code will use the broken version of `iter_matches`. By deprecating `iter_matches` in favor of `iter_matches2` in our documentation, anyone searching for docs on `iter_matches` will be easily and obviously directed to use the newer version instead. In contrast, an option is much less obvious and much easier to miss.

I suppose we could make the option VERY OBVIOUS in the docs. Perhaps something like this:

````
--- Iterates the matches of self on a given range.
---
--- Iterate over all matches within a {node}. The arguments are the same as for
--- |Query:iter_captures()| but the iterated values are different: an (1-based)
--- index of the pattern in the query, a table mapping capture indices to a list
--- of nodes, and metadata from any directives processing the match.
---
--- You MUST set the "correct" option for this function to behave correctly!
--- In older versions of Nvim this function has incorrect behavior: the correct
--- behavior must be opted into for backward compatibility.
---
--- Example:
---
--- ```lua
--- for pattern, match, metadata in cquery:iter_matches(tree:root(), bufnr, 0, -1, { correct = true}) do
---   for id, nodes in pairs(match) do
---     local name = query.captures[id]
---     for _, node in ipairs(nodes) do
---       -- `node` was captured by the `name` capture in the match
---
---       local node_data = metadata[id] -- Node level metadata
---       ... use the info here ...
---     end
---   end
--- end
--- ```
```

Reply:
In any case, I committed the option variant in c9c398693d8057c5cd244b97012a52243654097e. Easy to drop that commit if we need to.

Reply:
Sounds good! I would say I slightly prefer some meaningful name (e.g. `as_list` or other candidates suggested by @clason), but no objections to going with an *obnoxious* name.

Reply:
I still prefer a more descriptive name (_what_ does `correct` correct?), and the warning text can be more concise (but still very explicit). I can make a concrete suggestion, if you want.

Reply:
(I see it's already more concise in the PR; I'll still make a suggestion to reformulate slightly as it's missing the deprecation notice.)

Reply:
Concrete suggestion welcome. Whatever gets this thing out the door 🫠